chalk-engine/src/logic.rs

```diff
@@ -1328,11 +1314,23 @@ impl<'forest, C: Context + 'forest, CO: ContextOps<C> + 'forest> SolveState<'for
             return None;
         }
 
+        let table = self.stack.top().table;
+        let table_goal = &self.forest.tables[table].table_goal;
+
+        let filtered_delayed_subgoals = delayed_subgoals
+            .into_iter()
+            .filter(|delayed_subgoal| {
+                let (canonicalized, _) =
+                    infer.fully_canonicalize_goal(self.context.interner(), delayed_subgoal);
```

Comment:
Hmm, so each of these is getting canonicalized twice (here and in `canonicalize_answer_subst` below. I wonder if it's worth creating a new `AnswerSubst` below by filtering the delayed subgoals. Though it might be a little less clean. Maybe leave a FIXME here for a missed optimization.

Reply:
Do you think you can give an example. We want to create the `C::CanonicalAnswerSubst` without actually calling  `canonicalize_answer_subst` or ?

Reply:
No, we would just do the filtering using the canonicalized answer subst created below. But it would be less clean right now, since we would essentially have to have a way to extract out those delayed subgoals, compare, and add back. But that would have to be through the `Context` trait, since `chalk-engine` can't access them directly. We *might* have all the functions around already (I don't remember), but this is fine for now.